### PR TITLE
[Model Runner v2] Oracle for model runner v2 - qwen3 dense model by default [1/N]

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

Oracle for model runner v2 - dense model by default 

Now the env function:

- Not set: using our oracle
- set to 1: force v2
- set to 0: force v1

We are testing "Qwen/Qwen3-0.6B" and "facebook/opt-125m" since they cover the most current v1 unit test.

Should land after https://github.com/vllm-project/vllm/pull/39353

## Test

Covered in unit test

### CODE REVIEW

File: vllm/config/vllm.py [hunk 1]
```diff
@@ -458,6 +465,17 @@ def num_speculative_tokens(self) -> int:
             return self.speculative_config.num_speculative_tokens
         return 0
 
+    @property
```
Comment (njhill) [inline vllm/config/vllm.py:495]: should make this cached?

Comment (yewentao256) [reply vllm/config/vllm.py:495]: I think keeping it as a plain property is fine for now. It is only read a handful of times during setup.

Using cached_property here would not buy much and would make the env-sensitive behavior slightly less explicit.

Comment (njhill) [reply vllm/config/vllm.py:495]: Makes sense, I was thinking if this property is ever used at runtime, but it's true that we shouldn't need to check it in more places than we do already.

File: vllm/config/vllm.py [hunk 2]
```diff
@@ -63,6 +63,13 @@
 
 logger = init_logger(__name__)
 
+DEFAULT_V2_MODEL_RUNNER_WHITELIST = frozenset(
+    {
+        "Qwen/Qwen3-0.6B",
+        # "facebook/opt-125m",  # a lot of unit tests assert v1 specific feature
```
Comment (njhill) [inline vllm/config/vllm.py]: > `# a lot of unit tests assert v1 specific feature`

Actually should be a good reason to include it here? So that we ensure the oracle is falling back to v1 in all currently-unsupported cases.

Comment (yewentao256) [reply vllm/config/vllm.py:70]: "facebook/opt-125m" involves too many changes to cover, I prefer split it into a second PR, what do you think?

**Review by yewentao256 — COMMENTED** [review]
CC @njhill if we could land this PR first, [buildkite/ci/pr/cpu-language-generation-and-pooling-model-tests](https://buildkite.com/vllm/ci/builds/60886#019d79d4-8b1f-4e23-abe2-038c486497c4) is an unrelated CI failure.

Then I will have another PR adding opt 125M model

Also CC @WoosukKwon

**Review by njhill — COMMENTED** [review]
Thanks @yewentao256!

The changes to env var handling look good, but I'm not sure about the changes to various test files to set `VLLM_USE_V2_MODEL_RUNNER=0`. The point of the oracle is that this should be done automatically in cases where MRV2 doesn't yet work?

**Comment by yewentao256** [comment]
> Thanks @yewentao256!
> 
> The changes to env var handling look good, but I'm not sure about the changes to various test files to set `VLLM_USE_V2_MODEL_RUNNER=0`. The point of the oracle is that this should be done automatically in cases where MRV2 doesn't yet work?

There are two cases:

1. some unit tests broken because issues in MRv2 or current code, they are fixed like https://github.com/vllm-project/vllm/pull/39353
2. some unit tests testing specific logic in MRv1, which we don't need to support it (now or forever), then we mark for `VLLM_USE_V2_MODEL_RUNNER=0` and skip.

**Comment by njhill** [comment]
@yewentao256 sorry I'm still not following..

> * some unit tests broken because issues in MRv2 or current code, they are fixed like [[Model Runner V2] Fix flex attention kv blocks calculation issue #39353](https://github.com/vllm-project/vllm/pull/39353)

That's N/A now right... because it's already merged? If there are other things broken in v2 then we should either fix those or change oracle logic temporarily to avoid v2 in those cases.

> * some unit tests testing specific logic in MRv1, which we don't need to support it (now or forever), then we mark for `VLLM_USE_V2_MODEL_RUNNER=0` and skip.

What specific logic is this? If it is some feature that we have decided to deprecate, we should add the appropriate comments/warnings, and update the oracle logic to fall back to v1 for that feature.

For example, the oracle should currently be falling back to v1 for ngram spec decoding (and log appropriate warning), rather than disabling mrv2 in the test https://github.com/vllm-project/vllm/pull/39337/changes#diff-d9e2bab3c0cf3957bdcca3f64e4eda7460731056cba6743c75c6410dd0a9dcbe

**Comment by yewentao256** [comment]
@njhill I see, refactored the code, and found another issue to fix in https://github.com/vllm-project/vllm/pull/39937

Thanks for the suggestion!

**Review by yewentao256 — COMMENTED** [review]
Hi @njhill OK, ready for review now, the current CI failure is not related

File: vllm/config/vllm.py (hunk 2 above)
Comment (njhill) [reply vllm/config/vllm.py]: @yewentao256 if the oracle is working then it shouldn't involve more changes, since the relevant tests will just fall back to mrv1 for those? Again I think this would be a good reason to include it...

File: vllm/config/vllm.py [hunk 3]
```diff
@@ -64,6 +64,13 @@
 
 logger = init_logger(__name__)
 
+DEFAULT_V2_MODEL_RUNNER_WHITELIST = frozenset(
```
Comment (njhill) [inline vllm/config/vllm.py]: Maybe rename?
```suggestion
DEFAULT_V2_MODEL_RUNNER_MODELS = frozenset(
```

File: vllm/config/vllm.py [hunk 4]
```diff
@@ -472,6 +479,36 @@ def num_speculative_tokens(self) -> int:
...
+            logger.warning_once(
+                "Model runner v2 does not yet support reasoning budget "
+                "enforcement; using the v1 model runner instead when "
+                "reasoning parsing is configured."
+            )
+            return False
+
+        return (
+            self.model_config is not None
+            and self.model_config.model in DEFAULT_V2_MODEL_RUNNER_WHITELIST
```
Comment (njhill) [inline vllm/config/vllm.py]: Maybe do this check at the top since it's likely to be False in most cases?

File: vllm/config/vllm.py [hunk 5]
```diff
@@ -472,6 +479,36 @@ def num_speculative_tokens(self) -> int:
...
+
+        if self.reasoning_config is not None:
+            # TODO: add reasoning budget enforcement to ModelRunnerV2.
+            logger.warning_once(
+                "Model runner v2 does not yet support reasoning budget "
+                "enforcement; using the v1 model runner instead when "
+                "reasoning parsing is configured."
+            )
+            return False
+
```
Comment (njhill) [inline vllm/config/vllm.py:509]: I think there are other things we need to check here?

In particular yet-to-be implemented features from https://docs.google.com/document/d/1gFqtDkcoqhy9j-X0ndshzbhapX1uNey1-wBENwGPI80/edit?tab=t.192m4u5k37xt#heading=h.i84xcin8owkj such as custom logits processors, etc.

**Review by njhill — COMMENTED** [review]
Thanks @yewentao256 it looks good! but I think some additional checks are needed?

File: vllm/config/vllm.py (hunk 5 above)
Comment (yewentao256) [reply vllm/config/vllm.py:509]: Added

File: vllm/config/vllm.py (hunk 4 above)
Comment (yewentao256) [reply vllm/config/vllm.py]: Fixed

File: vllm/config/vllm.py [hunk 6]
```diff
@@ -472,6 +479,29 @@ def num_speculative_tokens(self) -> int:
...
+        use_v2_model_runner = envs.VLLM_USE_V2_MODEL_RUNNER
+        if use_v2_model_runner is not None:
+            return use_v2_model_runner
+
+        if (
+            self.model_config is None
+            or self.model_config.model not in DEFAULT_V2_MODEL_RUNNER_MODELS
+        ):
+            return False
+
```
Comment (njhill) [inline vllm/config/vllm.py:494]: We should add a check of `from vllm.triton_utils import HAS_TRITON` here too... and return False if that is False (e.g. CPU case).

Comment (njhill) [reply vllm/config/vllm.py:494]: Actually that should also be checked in the `_validate_v2_model_runner` path

**Review by njhill — COMMENTED** [review]
Thanks @yewentao256 this looks much better! Just one other thing that I thought of.

File: vllm/config/vllm.py (hunk 6 above)
Comment (yewentao256) [reply vllm/config/vllm.py:494]: Fixed

File: vllm/config/vllm.py [hunk 7]
```diff
@@ -1797,35 +1834,72 @@ def __str__(self):
...
 
-        if self.model_config is not None and self.model_config.has_inner_state:
+        if model_config is not None and model_config.has_inner_state:
             unsupported.append("hybrid/mamba models")
 
         if self.parallel_config.prefill_context_parallel_size > 1:
             unsupported.append("prefill context parallelism")
 
+        if not HAS_TRITON:
+            unsupported.append("Triton is unavailable")
```
Comment (njhill) [inline vllm/config/vllm.py]: If we check here, then we don't need the other triton check above.

However, I thought it might be better to have separate checks on both paths (oracle and validate) and not check in this method, since it's not checking for user-specified unsupported feature, rather it's checking platform compatibility. The info/error message in each of these cases can then be clearer / more specific.

**Comment by njhill** [comment]
Thanks @yewentao256!

It looks like some of the test failures now might be exposing legitimate MRV2 bugs and/or oracle selection issues, which is good! We'll need to investigate each of them.

File: vllm/config/vllm.py (hunk 7 above)
Comment (yewentao256) [reply vllm/config/vllm.py]: Fixed

**Review by yewentao256 — COMMENTED** [review]
Remove `[ready-run-all-tests]` label as it is very expensive, note that we already pass the tests

**Review by ZJY0516 — COMMENTED** [review]
Can we also test some multimodal models?

**Comment by njhill** [comment]
@ZJY0516 yes please try out multimodal models with MRV2! This PR is just making it default for qwen3 dense LMs initially.

**Comment by ZJY0516** [comment]
> @ZJY0516 yes please try out multimodal models with MRV2! This PR is just making it default for qwen3 dense LMs initially.

Oh I see, qwen 3 vl is `Qwen3VLForConditionalGeneration` instead of `Qwen3ForCausalLM`

**Review by NickLucche — COMMENTED** [review]
Is the plan to do qwen->dense->moe->mla/dsa->mamba and progressing?

If we want to skip dense as I understand the need for moe (minus unsupported feature) will arise quickly if things go well, we may want to consider starting more aggressively here with all dense - or at least include llama, mistral etc

**Review by yewentao256 — COMMENTED** [review]
Thanks @NickLucche ! We can have a following up PR for all dense later, let's land this qwen one first as it's been quite a long time fixing CI related stuff, not sure if new CI issues will come up for other dense models, so this basic infrastructure is needed first.

**Comment by njhill** [comment]
Now just waiting for https://github.com/vllm-project/vllm/pull/42538.

**Review by njhill — APPROVED** [review]
Thanks @yewentao256!

**Comment by NickLucche** [comment]
In hindsight we might have wanted a #ready-run-all-tests label here

**Comment by njhill** [comment]
> In hindsight we might have wanted a #ready-run-all-tests label here

I did set it and the tests had passed, but we should have added it again after the final changes before merging. I'm not sure why it didn't catch the P/D issue though.
